### ISSUE TITLE
Encoded latitude and longitude being read as unsigned int

### ISSUE THREAD

**Issue by stephenj079** [opensounder/python-sllib#3]
Hello,

I have been using this great library to view sl2 data.  I was noticing an issue with the output coordinates.  The corrected latitude was coming out at 90 degrees, whereas the latitude output from SonarViewer was coming out correctly (~ -31.94). I tested modifying the F1/F3_FRAME type for 'lat_enc' from an unsigned int to a signed int ('i'), and this fixed the issue. 

I presume that the issue is that testing has only been done using data acquired in the northern hemisphere.  Let me know if you would like some test data showing the issue.

Kind regards,

Steve

**Comment by kmpm** [comment]
Hi,
I would love some test data from southern hemisphere. Since I live not that far from the polar circle (~600km) I'm not even close.
Mail link to file or the actual file to code [at] kmpm {dot} se 
A small-ish one please. 
Alternatively you can add it as a pull-request to the project https://github.com/opensounder/sample-data-lowrance
But yet again, a small-ish one, less than 50MB preferably.

**Comment by NicholasBallard** [comment]
@stephenj079 great job figuring that out with the encoding. Anything helps filling in the holes in our understanding of the different values in the frame.

Have you made any sense of the sounding data packet in each frame? It looks like a transposed "water column" where each value is a return ping from the sounders. Does this have to be interpolated in a three-dimensional cone to create an image of the topography? I am completely lost at this point in interpreting the data.

Best,
Nick

**Comment by stephenj079** [comment]
Hello,

I have attached a small sl2 file to this message.  Please note that the start and end of this file contains data acquired on land. Also, the water is very shallow (around 1 m), with a very soft bottom (sludge, reeds, etc). The data was acquired on Lake Joondalup in Perth, Western Australia, during a test of an HDS5 Lowrance system. 
[Sonar0010.zip](https://github.com/opensounder/python-sllib/files/5442107/Sonar0010.zip)

I think I have figured out the sounding data packet. I am assuming that it is a 1D profile of the received 'ping' energy, though I am only using a simple single source/receiver HDS5 sonar. I was under the impression that the vertical axis of each data sounding would be in time from ping to receiving, however it appears to be a sample interval in depth, that is a ratio of the 'lower_limit' parameter.  Very strange, coming from a seismic background!  

Below is a function to read the frame 'metadata' (for lack of better word) and the packet, and output them as a list of dicts and a numpy array, respectively.  You can probably tell that this is a modification of the function given in the examples of this library. Probably not the best way to output the data (might be better to treat each packet as an array attached to the corresponding frame? Probably something for you guys to decide if you want to use this code), but gives an idea of how the data packet can be read:

```python
def read_echogram(input_file):
    import sllib
    with open(input_file, "rb") as f:
        reader = sllib.Reader(f)
        header = reader.header
        print(header)
        data = []
        for frame in reader:
            data.append(frame.to_dict())
            # the code to unpack the data packet
            temp = struct.unpack("i" * (frame.packetsize // 4), frame.packet)
            temp = np.array(temp)
            try:
                echogram = np.vstack((echogram, temp))
            except NameError:
                echogram = temp
    echogram = echogram.T.astype("float32")
    return data, echogram
```

Here's a function to read an sl2 file (using above function) and then plot the data packet numpy array as a matplotlib imshow, and overlay the automated water bottom pick. Comparing the output to this to what is shown in the software Sonar Viewer for the same file shows they are the same, which is what I was aiming for. But if you see any discrepancies, please let me know! 

```python
def test_overlay_wb_echo(filename):
    import matplotlib.pyplot as plt
    # Load sounding data and echogram.
    data, echo = read_echogram(filename)
    ## Calculate sampling rate, using assumption that y-axis is in depth (ft)
    ## - not actually used in this function but given here as an illustration. 
    # time1 = np.array([i["time1"] for i in data])
    # dz = time1[1:] - time1[:-1]
    ## Calculate the water depth in y-axis samples
    wb = np.array([i["water_depth"] for i in data])
    lower_limit = np.array([i["lower_limit"] for i in data])
    wb_s =  wb * echo.shape[0] / lower_limit
    fig, axes = plt.subplots()
    axes.imshow(echo)
    axes.plot([i["frame_index"] for i in data],
                 wb_s, linewidth=2, color="red")
    fig.show()
```

I hope that helps.  The code isn't the best, as I've been rushing to get it working for a job, so please excuse anything that looks bad.  For example the "try / except" in the read function is pretty shoddy, but it meant I got the code working, fast. 

Looking forward to some feedback. 

Kind regards,
Steve

**Comment by kmpm** [comment]
Thanks!!!!
- Your lon/lat fixes is now released as v0.0.3.
- Testdata is included in the https://github.com/opensounder/sample-data-lowrance project
- The changes to signed integers is included in the https://github.com/opensounder/sounder-log-formats project

**Comment by NicholasBallard** [comment]
@stephenj079 Thanks for your work, this is fantastic.
